File: android/src/main/java/com/lightningkite/khrysalis/observables/ObservableProperty.ext.shared.kt
package com.lightningkite.khrysalis.observables

import com.lightningkite.khrysalis.*
import com.lightningkite.khrysalis.rx.addWeak
import com.lightningkite.khrysalis.rx.forever
import io.reactivex.Observable
import io.reactivex.annotations.CheckReturnValue
import io.reactivex.annotations.SchedulerSupport
import io.reactivex.disposables.Disposable
import io.reactivex.rxkotlin.subscribeBy

val <T> ObservableProperty<T>.observable: Observable<Box<T>> get() = onChange.startWith { it.onNext(boxWrap(value)) }
val <T> ObservableProperty<T>.observableNN: Observable<T> get() = onChange.startWith { it.onNext(boxWrap(value)) }.map { it -> it.value }
val <T> ObservableProperty<T>.onChangeNN: Observable<T> get() = onChange.map { it -> it.value }

@CheckReturnValue
inline fun <T> ObservableProperty<T>.subscribeBy(
    noinline onError: @Escaping() (Throwable) -> Unit = { it -> it.printStackTrace() },
    noinline onComplete: @Escaping() () -> Unit = {},
    crossinline onNext: @Escaping() (T) -> Unit = { it -> }
): Disposable = this.observable.subscribeBy(
    onError = onError,
    onComplete = onComplete,
    onNext = { boxed -> onNext(boxed.value) }
)

fun <E> includes(collection: MutableObservableProperty<Set<E>>, element: E): MutableObservableProperty<Boolean> {
    return collection.map { it ->
        it.contains(element)
    }.withWrite { it ->
        if (it) {
            collection.value = collection.value.plus(element)
        } else {
            collection.value = collection.value.minus(element)
        }
    }
}

fun ObservableProperty<Boolean>.whileActive(action: @Escaping() () -> Disposable): Disposable {
    var current: Disposable? = null
    return this.subscribeBy {
        if (it) {
            if (current == null) {
                current = action()
            }
        } else {
            current?.dispose()
            current = null
        }
    }
}

File: ios/Khrysalis/com/lightningkite/khrysalis/observables/ObservableProperty.ext.shared.swift
//Package: com.lightningkite.khrysalis.observables
//Converted using Khrysalis2

import Foundation
import RxSwift
import RxRelay



extension ObservableProperty {
    public var observable: Observable<T> {
        get {
            return onChange.startWith(value)
        }
    }
}
 

extension ObservableProperty {
    public var observableNN: Observable<T> {
        get {
            return onChange.startWith(value)
        }
    }
}
 

extension ObservableProperty {
    public var onChangeNN: Observable<T> {
        get {
            return onChange
        }
    }
}
 

extension ObservableProperty {
     public func subscribeBy(onError: @escaping (Throwable) -> Void = { (it) in 
        it.printStackTrace()
    }, onComplete: @escaping () -> Void = { () in 
    }, onNext: @escaping (T) -> Void = { (it) in 
    }) -> Disposable {
        return self.observable.subscribeBy(onError: onError, onComplete: onComplete, onNext: { (boxed) in 
            onNext(boxed.value)
        })
    }
     public func subscribeBy(_ onError: @escaping (Throwable) -> Void, _ onComplete: @escaping () -> Void = { () in 
    }, _ onNext: @escaping (T) -> Void = { (it) in 
    }) -> Disposable {
        return subscribeBy(onError: onError, onComplete: onComplete, onNext: onNext)
    }
}
 
 

public func includes<E>(collection: MutableObservableProperty<Set<E>>, element: E) -> MutableObservableProperty<Bool> {
    return collection.map{ (it) in 
        it.contains(element)
    }.withWrite{ (it) in 
        if it {
            collection.value = collection.value.plus(element)
        } else {
            collection.value = collection.value.minus(element)
        }
    }
}
public func includes<E>(_ collection: MutableObservableProperty<Set<E>>, _ element: E) -> MutableObservableProperty<Bool> {
    return includes(collection: collection, element: element)
}
 
